### PR TITLE
Merging backends (Review Very Carefully!)

### PR DESCRIPTION
**Pull request by xubosen**
- Merged conflicting implementations of the backend
- Moved files into different folders for clarity (packages)
- Put location factories into a single class
- Made builder class for in memory data access
- See more details in commit messages

### CODE REVIEW

**Review by yuxi229 — COMMENTED** [review]
please see my comments

File: src/main/data_access/InMemoryLocationDao.java [hunk 1]
```diff
@@ -68,22 +37,25 @@ public boolean idExists(String id) {
     }
 
     @Override
-    public AbstractLocation getLocation(String id) {
+    public Location getLocation(String id) {
         return locationMap.get(id);
     }
 
     @Override
     public Room getRoom(String roomCode) {
```
[the remark this answers is not in the document]
Comment (xubosen) [reply src/main/data_access/InMemoryLocationDao.java:45]: Will be handled when implementing the room exist use case

File: src/main/api_infrastructure/APIClientImpl.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+public class APIClientImpl implements APIClient {
+    private final String baseUrl;
+    private final Gson gson;
+
+    public APIClientImpl(String baseUrl) {
+        this.baseUrl = baseUrl;
+        this.gson = new Gson(); // Using Gson for JSON parsing
+    }
+
+    private Map<String, Object> fetchData(String endpoint) {
```
[the remark this answers is not in the document]
Comment (xubosen) [reply src/main/api_infrastructure/APIClientImpl.java:21]: This is an issue! For clean architecture purposes, Map<String,Object> should be wrapped in some kind of proxy object. Needs to be worked on further;

File: src/main/entity/AbstractLocation.java [hunk 3]
```diff
@@ -5,26 +5,45 @@
...
         this.id = id;
+        this.connectedLocations = connectedLocations;
+        this.size = size;
     }
 
     public String getId() {
         return id;
     }
 
-    /**
```
[the remark this answers is not in the document]
Comment (xubosen) [reply src/main/entity/AbstractLocation.java:19]: This will be implemented differently in child classes!

**Review by yuxi229 — APPROVED** [review]

**Pull request merged**